### PR TITLE
PSI: add apply* functions that use ApplyPlanningScene.srv

### PR DESCRIPTION
**Pull request by v4hn**
This implements my comment https://github.com/ros-planning/moveit/pull/128#issuecomment-259990241
and supersedes \#128.

### CODE REVIEW

File: moveit_ros/planning_interface/planning_scene_interface/include/moveit/planning_scene_interface/planning_scene_interface.h [hunk 1]
```diff
@@ -92,11 +93,32 @@ class PlanningSceneInterface
...
+  bool applyCollisionObjects(const std::vector<moveit_msgs::CollisionObject> &collision_objects);
+
+  /** \brief Apply attached collision object to the planning scene of the move_group node.
+      Other PlanningSceneMonitors will NOT receive the update unless they subscribe to move_group's monitored scene */
+  bool applyAttachedCollisionObject(const moveit_msgs::AttachedCollisionObject &attached_collision_objects);
+
+  /** \brief Apply attached collision objects to the planning scene of the move_group node.
+      Other PlanningSceneMonitors will NOT receive the update unless they subscribe to move_group's monitored scene */
+  bool
+  applyAttachedCollisionObjects(const std::vector<moveit_msgs::AttachedCollisionObject> &attached_collision_objects);
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/planning_scene_interface/include/moveit/planning_scene_interface/planning_scene_interface.h]: what do you think of abbreviating ``collision_objects`` to ``collision_objs`` for line length?

File: moveit_ros/planning_interface/planning_scene_interface/include/moveit/planning_scene_interface/planning_scene_interface.h [hunk 2]
```diff
@@ -92,11 +93,32 @@ class PlanningSceneInterface
...
+
+  /** \brief Update the planning_scene of the move_group node with the given ps.
+      Other PlanningSceneMonitors will NOT receive the update unless they subscribe to move_group's monitored scene */
+  bool applyPlanningScene(const moveit_msgs::PlanningScene &ps);
+
+  /** \brief Add collision objects to the world via /planning_scene.
       Make sure object.operation is set to object.ADD. */
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/planning_scene_interface/include/moveit/planning_scene_interface/planning_scene_interface.h]: Add comment that this and the following function are the *old* asynchronous approach that usually require arbitrary sleep timers?

File: moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp [hunk 3]
```diff
@@ -210,6 +213,19 @@ class PlanningSceneInterface::PlanningSceneInterfaceImpl
...
   }
 
+  bool applyPlanningScene(const moveit_msgs::PlanningScene &planning_scene) const
+  {
+    moveit_msgs::ApplyPlanningScene::Request request;
+    moveit_msgs::ApplyPlanningScene::Response response;
+    request.scene = planning_scene;
+    if (!apply_planning_scene_service_.call(request, response))
+    {
+      ROS_WARN("Failed to call ApplyPlanningScene service");
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp]: ``_NAMESPACE``

File: moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp [hunk 4]
```diff
@@ -235,6 +251,7 @@ class PlanningSceneInterface::PlanningSceneInterfaceImpl
 private:
   ros::NodeHandle node_handle_;
   ros::ServiceClient planning_scene_service_;
+  mutable ros::ServiceClient apply_planning_scene_service_;
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp]: I know ``mutable`` is for ABI, but it seems dirty and we have sonames that will force everyone to recompile anyway. I vote to remove this.

**Comment by v4hn** [comment]
On Thu, Dec 15, 2016 at 11:35:02AM -0800, Dave Coleman wrote:
> what do you think of abbreviating ``collision_objects`` to ``collision_objs`` for line length?

I kept it the same as in the other functions and I would like to keep it that way.

> Add comment that this and the following function are the *old* asynchronous approach that usually require arbitrary sleep timers?

Well, they are not "old" as in "obsolete". I added comments on the synchronous/asynchronous behavior.

> ``_NAMESPACE``

done. `_NAMED` was missing throughout the class.

> I know ``mutable`` is for ABI, but it seems dirty and we have sonames that will force everyone to recompile anyway. I vote to remove this.

Actually, this is not at all required anymore, I just forgot to remove it... Thanks for pointing it out!

File: moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp [hunk 5]
```diff
@@ -210,6 +214,19 @@ class PlanningSceneInterface::PlanningSceneInterfaceImpl
     return result;
   }
 
+  bool applyPlanningScene(const moveit_msgs::PlanningScene &planning_scene) const
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp]: no ``const`` here? Travis is not building

Comment (v4hn) [inline moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp]: Yes, another neglected specifier that was shadowed by the `mutable` above.
I adjusted it. Let's see whether it passes now.

**Comment by davetcoleman** [comment]
Your comment in reference to Travis of "Let's see whether it passes now." worries me that this PR has not been tested on a live system at all. Is this ready to be merged?

**Comment by v4hn** [comment]
> Your comment in reference to Travis of "Let's see whether it passes now." worries me that this PR has not been tested on a live system at all. Is this ready to be merged?

Sorry to make you worry :)
I tested all new methods (before I opened this pr and just now) and the request is ready to merge.
I didn't test without the `mutable` before, because I didn't expect it would change anything to remove it. Turned out I missed something there..

I just rebased the request to clean up the history.

**Pull request merged**